### ISSUE TITLE
Decide if 'internal' JavaScript files should be exported from the package

### ISSUE THREAD

**Issue by 36degrees** [alphagov/govuk-frontend#3904]
## What

Decide if we want to restrict the files that we export (through the `exports` field in `package.json`, affecting which files can be resolved through package resolution) to only those we consider to be public.

## Why

The [NodeJS docs on the `exports` field](https://nodejs.org/api/packages.html#package-entry-points) suggest that it can be used to 'clearly define the public interface for [the] package' by only exporting the files that we consider to be public.

At the minute we export everything using `"./*"`, however there are only a number of entrypoints that we consider public (the top-level `all` and the individual components).

Everything in these files is currently private:

- `common/closest-attribute-value.mjs`
- `common/index.mjs`
- `common/normalise-dataset.mjs`
- `i18n.mjs` – we have plans to make it public in the future (https://github.com/alphagov/govuk-frontend/issues/2977)

And it's ambiguous whether this file is public or private:

- `common/govuk-frontend-version.mjs`

Removing files we consider internal / private from the exports will mean that they cannot be imported by consumers of the package – attempting to do so will result in [`ERR_PACKAGE_PATH_NOT_EXPORTED`](https://nodejs.org/api/errors.html#err_package_path_not_exported) error being thrown.

## Further details

This follows on from a discussion in https://github.com/alphagov/govuk-frontend/pull/3726#pullrequestreview-1514058496.

## Who needs to work on this

Developers

## Who needs to review this

Developers

## Done when

- [x] We've made a decision about whether to continue including internal modules in our package exports, or to restrict them to just those files we consider public
- [x] The `exports` field in `package.json` reflects that decision

**Comment by 36degrees** [comment]
Added to the agenda for next Monday's dev catch up

**Comment by 36degrees** [comment]
We discussed this in today's dev catch up.

We decided that for now we will [continue to export everything from our package](https://github.com/alphagov/govuk-frontend/blob/88642970c5a4fa2424036d9998185804dc8a43a8/packages/govuk-frontend/package.json#L21C18-L21C18), because:

- it's less likely that we'll break something, either now or in the future, by e.g. having mismatched exports to what we actually want to export
- it allows users to access our internal API items if they need to (either for use at risk or to patch something)
- the review app currently 'cheats' and imports things from the `src` directory, despite the `src` directory not being part of the published package – restricting the exports would prevent this and require us to make further changes to the review app

We will instead adopt [API Extractor's concept of release tags](https://api-extractor.com/pages/tsdoc/doc_comment_syntax/#release-tags) and [mark internal classes and functions as `@internal`](https://github.com/alphagov/govuk-frontend/issues/4017).

We can revisit this in the future if we find that we need to be stricter about what users can import from the published package.
